arangod/RestHandler/RestAdminClusterHandler.cpp

```diff
@@ -0,0 +1,1563 @@
+    return RestStatus::DONE;
+  }
+
+  auto const& suffixes = request()->suffixes();
+  size_t const len = suffixes.size();
+
+  if (len == 1) {
+    std::string const& command = suffixes.at(0);
+
+    if (command == Health) {
+      return handleHealth();
+    } else if (command == NumberOfServers) {
+      return handleNumberOfServers();
+    } else if (command == Maintenance) {
+      return handleMaintenance();
+    } else if (command == NodeVersion) {
+      return handleNodeVersion();
+    } else if (command == NodeEngine) {
+      return handleNodeEngine();
+    } else if (command == NodeStatistics) {
+      return handleNodeStatistics();
+    } else if (command == Statistics) {
+      return handleStatistics();
+    } else if (command == ShardDistribution) {
+      return handleShardDistribution();
+    } else if (command == CollectionShardDistribution) {
+      return handleCollectionShardDistribution();
+    } else if (command == CleanoutServer) {
+      return handleCleanoutServer();
+    } else if (command == ResignLeadership) {
+      return handleResignLeadership();
+    } else if (command == MoveShard) {
+      return handleMoveShard();
+    } else if (command == QueryJobStatus) {
+      return handleQueryJobStatus();
+    } else if (command == RemoveServer) {
+      return handleRemoveServer();
+    } else if (command == RebalanceShards) {
+      return handleRebalanceShards();
+    } else {
+      generateError(rest::ResponseCode::BAD, TRI_ERROR_HTTP_BAD_PARAMETER,
+                    std::string("invalid command '") + command + "'");
+      return RestStatus::DONE;
+    }
+  }
+
+  generateError(rest::ResponseCode::BAD, TRI_ERROR_HTTP_SUPERFLUOUS_SUFFICES,
+                "expecting URL /_admin/cluster/<command>");
+  return RestStatus::DONE;
+}
+
+
+RestAdminClusterHandler::futureVoid RestAdminClusterHandler::retryTryDeleteServer(std::unique_ptr<RemoveServerContext>&& ctx) {
+  if (++ctx->tries < 60) {
+    return SchedulerFeature::SCHEDULER->delay(1s).thenValue(
+      [this, ctx = std::move(ctx)] (auto) mutable {
+        return tryDeleteServer(std::move(ctx));
+    });
+  } else {
+    generateError(rest::ResponseCode::REQUEST_TIMEOUT, TRI_ERROR_HTTP_PRECONDITION_FAILED, "server may not be deleted");
+    return futures::makeFuture();
+  }
+}
+
+RestAdminClusterHandler::futureVoid RestAdminClusterHandler::tryDeleteServer(std::unique_ptr<RemoveServerContext>&& ctx) {
+
+  auto rootPath = arangodb::cluster::paths::root()->arango();
+  VPackBuffer<uint8_t> trx;
+  {
+    VPackBuilder builder(trx);
+    arangodb::agency::envelope<VPackBuilder>::create(builder)
+      .read()
+        .key(rootPath->supervision()->health()->str())
+        .key(rootPath->plan()->str())
+        .key(rootPath->current()->str())
+        .done()
+      .done();
+  }
+
+  return AsyncAgencyComm().sendWriteTransaction(20s, std::move(trx))
+    .thenValue([this, ctx = std::move(ctx)](AsyncAgencyCommResult &&result) mutable {
```

Comment:
```suggestion
    .thenValue([this, ctx = std::move(ctx)](AsyncAgencyCommResult&&result) mutable {
```
Do we need to capture `self` here as well?

Reply:
This lambda is executed (or destroyed) strictly before the `thenFinal` lambda in `waitForFuture` which captures `self`.